File: backend/src/core/signatures.py
"""
HMAC-SHA256 簽章驗證模組
用於 API 請求簽章生成與驗證
"""
import hmac
import hashlib
from typing import Dict, Any
from django.conf import settings


def generate_signature(params: Dict[str, Any], secret_key: str = None) -> str:
    """
    生成 HMAC-SHA256 簽章
    
    Args:
        params: 請求參數字典
        secret_key: API 密鑰（預設從 settings 讀取）
    
    Returns:
        小寫十六進位字串簽章
    """
    if secret_key is None:
        secret_key = settings.API_SECRET_KEY
    
    # 1. 參數排序（ASCII 順序）
    sorted_params = sorted(params.items())
    
    # 2. 生成查詢字串
    query_string = '&'.join([f"{k}={v}" for k, v in sorted_params if k != 'sign'])
    
    # 3. 加上 secret key
    query_string += f"&key={secret_key}"
    
    # 4. HMAC-SHA256
    signature = hmac.new(
        secret_key.encode('utf-8'),
        query_string.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    
    return signature


def verify_signature(params: Dict[str, Any], provided_sign: str, secret_key: str = None) -> bool:
    """
    驗證 HMAC-SHA256 簽章
    
    Args:
        params: 請求參數字典（包含 sign 和 timestamp）
        provided_sign: 提供的簽章
        secret_key: API 密鑰（預設從 settings 讀取）
    
    Returns:
        驗證是否通過
    """
    if 'sign' not in params:
        return False
    
    expected_sign = generate_signature(params, secret_key)
    
    # 使用 constant-time 比較避免時間攻擊
    return hmac.compare_digest(expected_sign, provided_sign)

File: frontend/src/lib/signatures.ts
/**
 * HMAC-SHA256 簽章生成工具
 */

const API_SECRET_KEY = process.env.NEXT_PUBLIC_API_SECRET_KEY || '';

/**
 * 生成 HMAC-SHA256 簽章
 * @param params 請求參數（不包含 sign）
 * @returns 簽章字串
 */
export async function generateSignature(
  params: Record<string, any>
): Promise<string> {
  // 使用 Web Crypto API 進行 HMAC-SHA256
  const encoder = new TextEncoder();
  
  // 1. 參數排序（ASCII 順序）
  const sortedParams = Object.entries(params)
    .filter(([key]) => key !== 'sign')
    .sort(([a], [b]) => a.localeCompare(b));
  
  // 2. 生成查詢字串
  const queryString = sortedParams
    .map(([key, value]) => `${key}=${value}`)
    .join('&');
  
  // 3. 加上 secret key
  const fullString = `${queryString}&key=${API_SECRET_KEY}`;
  
  // 4. HMAC-SHA256
  const keyData = encoder.encode(API_SECRET_KEY);
  const messageData = encoder.encode(fullString);
  
  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    keyData,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  
  const signatureBuffer = await crypto.subtle.sign(
    'HMAC',
    cryptoKey,
    messageData
  );
  
  // 轉換為十六進位字串
  const signatureArray = Array.from(new Uint8Array(signatureBuffer));
  const signatureHex = signatureArray
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
  
  return signatureHex;
}

/**
 * 為請求參數生成簽章並添加 timestamp
 */
export async function signRequest(
  params: Record<string, any>
): Promise<Record<string, any>> {
  const timestamp = Date.now();
  const paramsWithTimestamp = {
    ...params,
    timestamp,
  };
  
  const sign = await generateSignature(paramsWithTimestamp);
  
  return {
    ...paramsWithTimestamp,
    sign,
  };
}

